mitodl/ocw-studio
# mirror drive s3 upload

#### Pre-Flight checklist

  - [ ] Opened issue for DevOps regarding necessary configuration changes to deployed environments - https://github.com/mitodl/salt-ops/pull/1556

#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1689

#### What's this PR do?
This PR adds code to Concourse pipelines that, if an offline build is being run, uploads the output to a separate set of S3 buckets defined by `AWS_OFFLINE_PREVIEW_BUCKET_NAME` and `AWS_OFFLINE_PUBLISH_BUCKET_NAME`. This is being done as part of https://github.com/mitodl/hq/issues/785 to support creating a mirror drive copy of the entire OCW site.

#### How should this be manually tested?
 - Make a temporary edit to `websites/views.py`, changing line 341 to `sites = sites.prefetch_related("starter").order_by("name")[:10]` to limit the mass build to 10 sites
 - Spin up `ocw-studio`, making sure you have Concourse and Minio configured and the following in your `.env`:
```
OCW_HUGO_THEMES_BRANCH=main
OCW_HUGO_PROJECTS_BRANCH=main
AWS_OFFLINE_PREVIEW_BUCKET_NAME=ocw-content-offline-draft
AWS_OFFLINE_PUBLISH_BUCKET_NAME=ocw-content-offline-live
```
 - Run `docker compose exec web ./manage.py upsert_theme_assets_pipeline` to ensure you have the latest version of the theme assets pipeline pushed up
 - Browse to the Concourse UI at http://localhost:3000 and go to the `ocw-theme-assets` group, finding the `main` branch version that we just pushed up
 - Unpause and trigger a build of the theme assets pipeline, waiting for it to complete
 - Run `docker compose exec web ./manage.py upsert_mass_build_pipeline --offline --starter ocw-course-v2`
 - Log into the Concourse UI at http://localhost:8080
 - Find the mass build pipeline you just pushed up with offline set to true, version set to live and starter set to ocw-course-v2
 - Trigger a build and wait for it to finish
 - Log into the Minio web interface at http://localhost:9001
 - Browse to `ocw-content-offline-live` and open the `courses` folder, then select a course folder, then click download in the right side context menu
 - Extract the ZIP file and double click on `index.html`
 - Double click on the `index.html` file to open the site and make sure:
   - There are no errors in the console
   - JS / CSS / Images load properly
   - Nav links work
   - Links to resources work
